File: tcpchat-client/internal/message.go
package internal

type EVENT string

const (
	RENAME    EVENT = "/name" // Client events
	JOIN_ROOM EVENT = "/join"
	GET_ROOMS EVENT = "/rooms"
	ERROR     EVENT = "/err" // Server event
	MESSAGE   EVENT = "/msg" // Share event
)

type message struct {
	Event   EVENT  `json:"event"`
	Payload string `json:"payload"`
}

File: tcpchat-server/src/server/message.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub enum Event {
    #[serde(rename = "/name")]
    Rename,
    #[serde(rename = "/join")]
    JoinRoom,
    #[serde(rename = "/rooms")]
    GetRooms,
    #[serde(rename = "/msg")]
    Message,
    #[serde(rename = "/err")]
    Error,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Message {
    pub event: Event,
    pub payload: String,
}

pub fn create_message(event: Event, payload: String) -> String {
    let mut data =
        serde_json::to_string(&Message { event, payload }).expect("Failed to serialize data");
    data.push('\n');
    data
}
